active/0014-rolling-cluster-upgrade.md

```diff
@@ -0,0 +1,122 @@
+
+## Motivation
+
+In order to make upgrading the cluster smoother, EMQ X should support rolling cluster upgrades and green-blue deployments.
+It should be possible to upgrade the cluster without taking it down entirely.
+
+## Design
+
+Live upgrade paths will be limited to `major.minor.* -> major.minor+1.*` formula.
+
+A new concept of a "backplane API" is introduced.
+
+Cluster upgrade should be split in roughly three stages:
+
+1. Optional step: inject the forward compatibility support into the old upgrade.
+   This may be needed because we can't predict in advance what will be changed in the next release.
+   The upgrade code should be able to inject pre-upgrade beams to the old release.
+   This part of the upgrade procedure should be idempotent and reversible.
+   It will work much like a patch version upgrade.
+1. Rolling upgrade of the cluster that involves taking nodes down and replacing them with the newer version.
+   Both versions of the code work together in the same cluster.
+   This part of the upgrade can be revesible.
+1. Once all the nodes are upgraded, a data migration can start.
+   This part of the upgrade is not reversible.
+
+Deprecated APIs can be removed in the next release.
+
+There are two major areas that need to be considered to support this kind of upgrade:
+
+1. RPC compatibility
+1. Mnesia schema backward-compatibility
+
+A helper mnesia table tracking the upgrade state and release version for each node can be used to perform the checks between proceeding to the next stage of the upgrade.
+
+### RPC compatibility
+
+In order to simplify the reasoning about the backplane API backward-compatibility, all the functions that may be called remotely should be identified and gathered in specialized modules.
+Static checks should be introduced to make sure these functions are not removed without being deprecated first.
+(Perhaps dump the list of these functions to a file that is held in the repo and used in the CI?)
+Direct sending messages to the remote processes should be prohibited.
+Instead, a helper function in the API module should be introduced.
+
+### Mnesia schema compatibility
+
+Fields of the tables can't be removed (until the last stage of the upgrade?).
+Non-trivial changes to the schema should be performed in stages:
+
+1. The new version of the code should be able to work with the old schema.
+1. Once all the nodes in the cluster are updated, start an async process of migrating the data to the new table.
+1. Once the data has been migrated and checks pass, the old table can be removed.
+1. In the next release the code supporting the old schema can be dropped.
+
+#### Table migration
+
+Writing migration in a way that avoids whole-table locks is a complicated process, and should be done on a case-by-case basis.
+Migration process could utilize mnesia transactions to traverse the tables entry-by-entry and move the records to the new table, deleting the old record.
```

Comment:
This is achieved by having different tables for different versions, right?

Reply:
Yes